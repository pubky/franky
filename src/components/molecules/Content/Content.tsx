import { ReactNode } from 'react';
import Image, { ImageProps } from 'next/image';

import * as Atoms from '@/atoms';
import * as Libs from '@/libs';

interface ContentCardProps {
  children?: ReactNode;
  className?: string;
  image?: {
    src: string;
    alt: string;
    width: number;
    height: number;
    size?: 'small' | 'medium' | 'large';
  };
  layout?: 'row' | 'column';
}

interface ContentContainerProps {
  children: ReactNode;
  className?: string;
  maxWidth?: 'sm' | 'md' | 'lg' | 'xl';
  gap?: 'sm' | 'md' | 'lg';
}

interface ContentImageProps extends Omit<ImageProps, 'className'> {
  className?: string;
  containerClassName?: string;
  size?: 'small' | 'medium' | 'large';
  hiddenOnMobile?: boolean;
}

export function ContentCard({ children, className, image, layout = 'row' }: ContentCardProps) {
  const layoutClasses = {
    row: 'flex-col lg:flex-row',
    column: 'flex-col',
  };

  return (
    <Atoms.Card className={Libs.cn('p-6 lg:p-12', className)}>
      <Atoms.Container className={Libs.cn('gap-12', layoutClasses[layout])}>
        {image && (
          <ContentImage src={image.src} alt={image.alt} width={image.width} height={image.height} size={image.size} />
        )}
        <Atoms.Container className="gap-6 justify-start w-full">{children}</Atoms.Container>
      </Atoms.Container>
    </Atoms.Card>
  );
}

export function ContentContainer({ children, className, maxWidth = 'lg', gap = 'md' }: ContentContainerProps) {
  const maxWidthClasses = {
    sm: 'max-w-[588px]',
    md: 'max-w-[800px]',
    lg: 'max-w-[1200px]',
    xl: 'max-w-[1400px]',
  };

  const gapClasses = {
    sm: 'gap-3',
    md: 'gap-6',
    lg: 'gap-8',
  };

  return (
    <Atoms.Container className={Libs.cn(maxWidthClasses[maxWidth], gapClasses[gap], className)}>
      {children}
    </Atoms.Container>
  );
}

export function ContentImage({
  className,
  containerClassName,
  size = 'medium',
  hiddenOnMobile = true,
  ...imageProps
}: ContentImageProps) {
  const sizeClasses = {
    small: 'w-full lg:w-[137px]',
    medium: 'w-full lg:w-[228px]',
    large: 'w-full lg:w-[265px]',
  };

  return (
    <div className={Libs.cn(sizeClasses[size], hiddenOnMobile ? 'hidden lg:flex' : 'flex', containerClassName)}>
      <Image
        data-testid="content-image"
        data-size={size}
        className={Libs.cn('w-full h-auto', className)}
        {...imageProps}
        alt={imageProps.alt || 'Image'}
      />
    </div>
  );
}
