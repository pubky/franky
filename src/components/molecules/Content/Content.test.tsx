/* eslint-disable @next/next/no-img-element */
import { describe, it, expect, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import { ContentCard, ContentContainer, ContentImage } from './Content';

// Mock Next.js Image component
vi.mock('next/image', () => ({
  default: ({
    src,
    alt,
    width,
    height,
    ...props
  }: {
    src: string;
    alt: string;
    width: number;
    height: number;
    [key: string]: unknown;
  }) => <img data-testid="next-image" src={src} alt={alt} width={width} height={height} {...props} />,
}));

// Mock atoms
vi.mock('@/atoms', () => ({
  Container: ({ children, className }: { children: React.ReactNode; className?: string }) => (
    <div data-testid="container" className={className}>
      {children}
    </div>
  ),
  Card: ({ children, className }: { children: React.ReactNode; className?: string }) => (
    <div data-testid="card" className={className}>
      {children}
    </div>
  ),
}));

describe('ContentCard', () => {
  it('renders with default props', () => {
    render(
      <ContentCard>
        <div>Test content</div>
      </ContentCard>,
    );

    expect(screen.getByTestId('card')).toBeInTheDocument();
    expect(screen.getByText('Test content')).toBeInTheDocument();
  });

  it('applies default classes to card', () => {
    render(
      <ContentCard>
        <div>Content</div>
      </ContentCard>,
    );

    const card = screen.getByTestId('card');
    expect(card).toHaveClass('p-6', 'md:p-12');
  });

  it('applies custom className to card', () => {
    render(
      <ContentCard className="custom-card">
        <div>Content</div>
      </ContentCard>,
    );

    const card = screen.getByTestId('card');
    expect(card).toHaveClass('custom-card');
  });

  it('renders with image', () => {
    const image = {
      src: '/test.jpg',
      alt: 'Test image',
      width: 200,
      height: 200,
    };

    render(
      <ContentCard image={image}>
        <div>Content with image</div>
      </ContentCard>,
    );

    const img = screen.getByTestId('content-image');
    expect(img).toBeInTheDocument();
    expect(img).toHaveAttribute('src', '/test.jpg');
    expect(img).toHaveAttribute('alt', 'Test image');
    expect(screen.getByText('Content with image')).toBeInTheDocument();
  });

  it('applies row layout by default', () => {
    render(
      <ContentCard>
        <div>Content</div>
      </ContentCard>,
    );

    const container = screen.getAllByTestId('container')[0];
    expect(container).toHaveClass('flex-col', 'lg:flex-row');
  });

  it('applies column layout when specified', () => {
    render(
      <ContentCard layout="column">
        <div>Content</div>
      </ContentCard>,
    );

    const container = screen.getAllByTestId('container')[0];
    expect(container).toHaveClass('flex-col');
    expect(container).not.toHaveClass('lg:flex-row');
  });
});

describe('ContentContainer', () => {
  it('renders with default props', () => {
    render(
      <ContentContainer>
        <div>Container content</div>
      </ContentContainer>,
    );

    expect(screen.getByTestId('container')).toBeInTheDocument();
    expect(screen.getByText('Container content')).toBeInTheDocument();
  });

  it('applies default max-width and gap', () => {
    render(
      <ContentContainer>
        <div>Content</div>
      </ContentContainer>,
    );

    const container = screen.getByTestId('container');
    expect(container).toHaveClass('max-w-[1200px]', 'gap-6');
  });

  it('applies custom max-width', () => {
    render(
      <ContentContainer maxWidth="sm">
        <div>Content</div>
      </ContentContainer>,
    );

    const container = screen.getByTestId('container');
    expect(container).toHaveClass('max-w-[588px]');
  });

  it('applies custom gap', () => {
    render(
      <ContentContainer gap="lg">
        <div>Content</div>
      </ContentContainer>,
    );

    const container = screen.getByTestId('container');
    expect(container).toHaveClass('gap-8');
  });

  it('applies custom className', () => {
    render(
      <ContentContainer className="custom-container">
        <div>Content</div>
      </ContentContainer>,
    );

    const container = screen.getByTestId('container');
    expect(container).toHaveClass('custom-container');
  });
});

describe('ContentImage', () => {
  it('renders with required props', () => {
    render(<ContentImage src="/test.jpg" alt="Test" width={100} height={100} />);

    const img = screen.getByTestId('content-image');
    expect(img).toBeInTheDocument();
    expect(img).toHaveAttribute('src', '/test.jpg');
    expect(img).toHaveAttribute('alt', 'Test');
    expect(img).toHaveAttribute('width', '100');
    expect(img).toHaveAttribute('height', '100');
  });

  it('hides on mobile by default', () => {
    const { container } = render(<ContentImage src="/test.jpg" alt="Test" width={100} height={100} />);

    const wrapper = container.firstChild as HTMLElement;
    expect(wrapper).toHaveClass('hidden', 'lg:flex');
  });

  it('shows on mobile when hiddenOnMobile is false', () => {
    const { container } = render(
      <ContentImage src="/test.jpg" alt="Test" width={100} height={100} hiddenOnMobile={false} />,
    );

    const wrapper = container.firstChild as HTMLElement;
    expect(wrapper).toHaveClass('flex');
    expect(wrapper).not.toHaveClass('hidden');
  });

  it('applies custom className to image', () => {
    render(<ContentImage src="/test.jpg" alt="Test" width={100} height={100} className="custom-image" />);

    const img = screen.getByTestId('content-image');
    expect(img).toHaveClass('custom-image');
  });

  it('applies container styling', () => {
    const { container } = render(
      <ContentImage src="/test.jpg" alt="Test" width={100} height={100} containerClassName="custom-container" />,
    );

    const wrapper = container.firstChild as HTMLElement;
    expect(wrapper).toHaveClass('custom-container');
    expect(wrapper.style.width).toBe('100px');
    expect(wrapper.style.height).toBe('100px');
  });
});

describe('Content - Snapshots', () => {
  it('matches snapshot for ContentCard with default props', () => {
    const { container } = render(
      <ContentCard>
        <div>Test content</div>
      </ContentCard>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshots for ContentCard with different configurations', () => {
    const { container: defaultContainer } = render(
      <ContentCard>
        <div>Default content</div>
      </ContentCard>,
    );
    expect(defaultContainer.firstChild).toMatchSnapshot();

    const { container: customClassContainer } = render(
      <ContentCard className="custom-card">
        <div>Custom content</div>
      </ContentCard>,
    );
    expect(customClassContainer.firstChild).toMatchSnapshot();

    const { container: columnLayoutContainer } = render(
      <ContentCard layout="column">
        <div>Column layout content</div>
      </ContentCard>,
    );
    expect(columnLayoutContainer.firstChild).toMatchSnapshot();

    const image = {
      src: '/test.jpg',
      alt: 'Test image',
      width: 200,
      height: 200,
    };
    const { container: withImageContainer } = render(
      <ContentCard image={image}>
        <div>Content with image</div>
      </ContentCard>,
    );
    expect(withImageContainer.firstChild).toMatchSnapshot();
  });

  it('matches snapshots for ContentContainer with different configurations', () => {
    const { container: defaultContainer } = render(
      <ContentContainer>
        <div>Default container</div>
      </ContentContainer>,
    );
    expect(defaultContainer.firstChild).toMatchSnapshot();

    const { container: customMaxWidthContainer } = render(
      <ContentContainer maxWidth="sm">
        <div>Small max width</div>
      </ContentContainer>,
    );
    expect(customMaxWidthContainer.firstChild).toMatchSnapshot();

    const { container: customGapContainer } = render(
      <ContentContainer gap="lg">
        <div>Large gap</div>
      </ContentContainer>,
    );
    expect(customGapContainer.firstChild).toMatchSnapshot();

    const { container: customClassContainer } = render(
      <ContentContainer className="custom-container">
        <div>Custom class</div>
      </ContentContainer>,
    );
    expect(customClassContainer.firstChild).toMatchSnapshot();
  });

  it('matches snapshots for ContentImage with different configurations', () => {
    const { container: defaultContainer } = render(
      <ContentImage src="/test.jpg" alt="Test" width={100} height={100} />,
    );
    expect(defaultContainer.firstChild).toMatchSnapshot();

    const { container: visibleOnMobileContainer } = render(
      <ContentImage src="/test.jpg" alt="Test" width={100} height={100} hiddenOnMobile={false} />,
    );
    expect(visibleOnMobileContainer.firstChild).toMatchSnapshot();

    const { container: customClassContainer } = render(
      <ContentImage src="/test.jpg" alt="Test" width={100} height={100} className="custom-image" />,
    );
    expect(customClassContainer.firstChild).toMatchSnapshot();

    const { container: customContainerClassContainer } = render(
      <ContentImage src="/test.jpg" alt="Test" width={100} height={100} containerClassName="custom-container" />,
    );
    expect(customContainerClassContainer.firstChild).toMatchSnapshot();
  });
});
