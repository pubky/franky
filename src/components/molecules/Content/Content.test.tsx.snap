// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Content - Snapshots > matches snapshot for ContentCard with column layout 1`] = `
<div
  class="p-6 md:p-12"
  data-testid="card"
>
  <div
    class="gap-12 flex-col"
    data-testid="container"
  >
    <div
      class="gap-3 justify-start w-full"
      data-testid="container"
    >
      <div>
        Column layout content
      </div>
    </div>
  </div>
</div>
`;

exports[`Content - Snapshots > matches snapshot for ContentCard with custom className 1`] = `
<div
  class="p-6 md:p-12 custom-card"
  data-testid="card"
>
  <div
    class="gap-12 flex-col lg:flex-row"
    data-testid="container"
  >
    <div
      class="gap-3 justify-start w-full"
      data-testid="container"
    >
      <div>
        Custom content
      </div>
    </div>
  </div>
</div>
`;

exports[`Content - Snapshots > matches snapshot for ContentCard with default props 1`] = `
<div
  class="p-6 md:p-12"
  data-testid="card"
>
  <div
    class="gap-12 flex-col lg:flex-row"
    data-testid="container"
  >
    <div
      class="gap-3 justify-start w-full"
      data-testid="container"
    >
      <div>
        Test content
      </div>
    </div>
  </div>
</div>
`;

exports[`Content - Snapshots > matches snapshot for ContentCard with image 1`] = `
<div
  class="p-6 md:p-12"
  data-testid="card"
>
  <div
    class="gap-12 flex-col lg:flex-row"
    data-testid="container"
  >
    <div
      class="hidden lg:flex"
      style="width: 200px; height: 200px;"
    >
      <img
        alt="Test image"
        class=""
        data-testid="content-image"
        height="200"
        src="/test.jpg"
        style="object-fit: contain;"
        width="200"
      />
    </div>
    <div
      class="gap-3 justify-start w-full"
      data-testid="container"
    >
      <div>
        Content with image
      </div>
    </div>
  </div>
</div>
`;

exports[`Content - Snapshots > matches snapshot for ContentContainer with custom className 1`] = `
<div
  class="max-w-[1200px] gap-6 custom-container"
  data-testid="container"
>
  <div>
    Custom class
  </div>
</div>
`;

exports[`Content - Snapshots > matches snapshot for ContentContainer with default props 1`] = `
<div
  class="max-w-[1200px] gap-6"
  data-testid="container"
>
  <div>
    Default container
  </div>
</div>
`;

exports[`Content - Snapshots > matches snapshot for ContentContainer with large gap 1`] = `
<div
  class="max-w-[1200px] gap-8"
  data-testid="container"
>
  <div>
    Large gap
  </div>
</div>
`;

exports[`Content - Snapshots > matches snapshot for ContentContainer with small max width 1`] = `
<div
  class="max-w-[588px] gap-6"
  data-testid="container"
>
  <div>
    Small max width
  </div>
</div>
`;

exports[`Content - Snapshots > matches snapshot for ContentImage visible on mobile 1`] = `
<div
  class="flex"
  style="width: 100px; height: 100px;"
>
  <img
    alt="Test"
    class=""
    data-testid="content-image"
    height="100"
    src="/test.jpg"
    style="object-fit: contain;"
    width="100"
  />
</div>
`;

exports[`Content - Snapshots > matches snapshot for ContentImage with custom className 1`] = `
<div
  class="hidden lg:flex"
  style="width: 100px; height: 100px;"
>
  <img
    alt="Test"
    class="custom-image"
    data-testid="content-image"
    height="100"
    src="/test.jpg"
    style="object-fit: contain;"
    width="100"
  />
</div>
`;

exports[`Content - Snapshots > matches snapshot for ContentImage with custom container className 1`] = `
<div
  class="hidden lg:flex custom-container"
  style="width: 100px; height: 100px;"
>
  <img
    alt="Test"
    class=""
    data-testid="content-image"
    height="100"
    src="/test.jpg"
    style="object-fit: contain;"
    width="100"
  />
</div>
`;

exports[`Content - Snapshots > matches snapshot for ContentImage with default props 1`] = `
<div
  class="hidden lg:flex"
  style="width: 100px; height: 100px;"
>
  <img
    alt="Test"
    class=""
    data-testid="content-image"
    height="100"
    src="/test.jpg"
    style="object-fit: contain;"
    width="100"
  />
</div>
`;
